Make CI error on inductor fallback when decomp is available

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #99473
* #99659

Fixes #99446 

Remove the warning, as that annoyed end-users who don't know what to do about it.

Instead, try to hold the line by preventing any decomp from being added without making
the corresponding change to inductor's fallbacks.

Note: we probably still need to better document how to update inductor's decomps,
for now it's pretty much "go ask the inductor team for advice"

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

wconstab committed:
Make CI error on inductor fallback when decomp is available

Remove the warning, as that annoyed end-users who don't know what to do about it.

Instead, try to hold the line by preventing any decomp from being added without making
the corresponding change to inductor's fallbacks.

Note: we probably still need to better document how to update inductor's decomps,
for now it's pretty much "go ask the inductor team for advice"

[ghstack-poisoned]

wconstab committed:
Update on "Make CI error on inductor fallback when decomp is available"


Fixes #99446 

Remove the warning, as that annoyed end-users who don't know what to do about it.

Instead, try to hold the line by preventing any decomp from being added without making
the corresponding change to inductor's fallbacks.

Note: we probably still need to better document how to update inductor's decomps,
for now it's pretty much "go ask the inductor team for advice"

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

ezyang commented on torch/_inductor/lowering.py:
```diff
@@ -1085,10 +1087,19 @@ def make_fallback(kernel, layout_constraint=None, warn=True):
         )
+        # Make sure the process exits and CI flags an error, raising
+        # an exception could allow a graph-break and silently pass CI in some configs
+        sys.exit(1)
```
This seems cure is worse than disease. Why don't we raise an exception and write off the suppress error configs?

wconstab replied:
I mean, I kinda knew you'd say that. But on balance, this way will stop fallbacks from getting checked in broken, and the only downside that I'm cognizant of is that it robs you of getting notified of further-down-the-stack failures on CI if it exits early.  I think it's slightly better to be sure not to allow bad things to land than to make devs rerun CI more times.  But ...

ezyang replied:
No, I don't understand your response. What does `sys.exit(0)` give you that `raise AssertionError()` wouldn't?

wconstab replied:
I am worried that dynamo will swallow exceptions sometimes in some tests in its attempt to gracefully fall back to eager. However, I hope there is some type of internal assertion we can raise that will reliably raise to userland, and if so i'm happy to use that instead.

jansel replied:
Perhaps we should do
```
if TEST_WITH_DYNAMO:
   sys.exit(1)
else:
   raise ....
```

Under `TEST_WITH_DYNAMO` raising will just fall back silently to eager and skip the real test.


Another option would be to dynamically switch `suppress_errors=False` and raise.

wconstab committed:
Update on "Make CI error on inductor fallback when decomp is available"


Fixes #99446 

Remove the warning, as that annoyed end-users who don't know what to do about it.

Instead, try to hold the line by preventing any decomp from being added without making
the corresponding change to inductor's fallbacks.

Note: we probably still need to better document how to update inductor's decomps,
for now it's pretty much "go ask the inductor team for advice"

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Make CI error on inductor fallback when decomp is available"


Fixes #99446 

Remove the warning, as that annoyed end-users who don't know what to do about it.

Instead, try to hold the line by preventing any decomp from being added without making
the corresponding change to inductor's fallbacks.

Note: we probably still need to better document how to update inductor's decomps,
for now it's pretty much "go ask the inductor team for advice"

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Make CI error on inductor fallback when decomp is available"


Fixes #99446 

Remove the warning, as that annoyed end-users who don't know what to do about it.

Instead, try to hold the line by preventing any decomp from being added without making
the corresponding change to inductor's fallbacks.

Note: we probably still need to better document how to update inductor's decomps,
for now it's pretty much "go ask the inductor team for advice"

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Make CI error on inductor fallback when decomp is available"


Fixes #99446 

Remove the warning, as that annoyed end-users who don't know what to do about it.

Instead, try to hold the line by preventing any decomp from being added without making
the corresponding change to inductor's fallbacks.

Note: we probably still need to better document how to update inductor's decomps,
for now it's pretty much "go ask the inductor team for advice"

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Make CI error on inductor fallback when decomp is available"


Fixes #99446 

Remove the warning, as that annoyed end-users who don't know what to do about it.

Instead, try to hold the line by preventing any decomp from being added without making
the corresponding change to inductor's fallbacks.

Note: we probably still need to better document how to update inductor's decomps,
for now it's pretty much "go ask the inductor team for advice"

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Make CI error on inductor fallback when decomp is available"


Fixes #99446 

Remove the warning, as that annoyed end-users who don't know what to do about it.

Instead, try to hold the line by preventing any decomp from being added without making
the corresponding change to inductor's fallbacks.

Note: we probably still need to better document how to update inductor's decomps,
for now it's pretty much "go ask the inductor team for advice"

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Make CI error on inductor fallback when decomp is available"


Fixes #99446 

Remove the warning, as that annoyed end-users who don't know what to do about it.

Instead, try to hold the line by preventing any decomp from being added without making
the corresponding change to inductor's fallbacks.

Note: we probably still need to better document how to update inductor's decomps,
for now it's pretty much "go ask the inductor team for advice"

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

wconstab commented:
@ngimel please take another look as I have taken the liberty of bypassing one decomp that i found to cause test failures in inductor when using.  If you're ok with this, I can open another issue to track the bad decomp.

ngimel reviewed (approved):
not using `exponential` decomp is fine, but why do you need separate DENY_LIST for it? Can you remove it from extra random decomps and make a fallback instead?

wconstab commented:
> not using exponential decomp is fine, but why do you need separate DENY_LIST for it? Can you remove it from extra random decomps and make a fallback instead?

Hm,  yes, why did I not do that. I guess i can make an explicit fallback with warn=False for that one.

wconstab committed:
Update on "Make CI error on inductor fallback when decomp is available"


Fixes #99446 

Remove the warning, as that annoyed end-users who don't know what to do about it.

Instead, try to hold the line by preventing any decomp from being added without making
the corresponding change to inductor's fallbacks.

Note: we probably still need to better document how to update inductor's decomps,
for now it's pretty much "go ask the inductor team for advice"

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

wconstab commented:
@pytorchbot merge
